File: code/timer2/recipe-02/examples/timer_once.cpp
#include "timer.hpp"
#include "print_message.hpp"
#include <chrono>
#include <thread>

void say_hello() {
    print_message("Hello, World!");
}

int main() {
    auto timer = Timer{say_hello, std::chrono::seconds{3}};
    print_message("Timer started, waiting for it to trigger...");
    while (timer.isRunning()) {
        print_message("Timer is running...");
        std::this_thread::sleep_for(std::chrono::seconds{1});
    }
    print_message("Timer is not running.");
}

File: code/threading.Timer2/python/timer_once.py
#!/usr/bin/env python3

import threading
import datetime
import time

def print_message(message):
    print(f"{datetime.datetime.now()}: {message}")

def say_hello():
    print_message("Hello, World!")

def main():
    timer = threading.Timer(3.0, say_hello)
    timer.start()
    
    print_message("Timer started, waiting for it to trigger...")
    while timer.is_alive():
        print_message("Timer is active...")
        time.sleep(1)
    print_message("Timer is not active.");

if __name__ == "__main__":
    main()
